pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -84,6 +95,13 @@ protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutab
       _primaryKeyToRecordLocationMap.compute(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction),
           (primaryKey, currentRecordLocation) -> {
             if (currentRecordLocation != null) {
+              // Skip the records that has comparisonValue timestamp earlier than (largestSeenTimestamp - TTL).
```

Comment:
I don't follow this logic. Do we need this check?

Reply:
@Jackie-Jiang I think this is optional. Do you suggest me to remove it?

For segments out of retention, we can skip them based on segmentMetadata.
For segments within retention, some records has comparison column value out of retention, this logic will skip these records.

Reply:
@Jackie-Jiang i will remove it for this PR for now. Comparing with numbers of segment that out-of-retention, the numbers of out-of-TTL records in valid segments (within TTL) are small.